File: PIDEV-YBrainy-E-leraning-certifications-Platform/payment/Payment/src/main/java/com/backend/service/PackConversionService.java
package com.backend.service;

import com.backend.dto.packconversion.PackConversionSummaryResponseDTO;
import com.backend.exception.BusinessRuleException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class PackConversionService {

    private final ObjectMapper objectMapper;

    @Value("${pack-conversion.output-directory:../Conversion packs/outputs}")
    private String packConversionOutputDirectory;

    @Value("${pack-conversion.summary-file-name:pack_conversion_summary.json}")
    private String packConversionSummaryFileName;

    public PackConversionSummaryResponseDTO getSummary(int limit) {
        if (limit <= 0) {
            throw new BusinessRuleException("Limit must be greater than zero.");
        }

        Path summaryPath = resolveSummaryPath();
        JsonNode root = readSummaryJson(summaryPath);
        List<PackConversionSummaryResponseDTO.PackConversionScoreDTO> topPacks =
                parseTopConversionPacks(root.path("top_conversion_packs"), limit);

        return new PackConversionSummaryResponseDTO(
                summaryPath.toString(),
                parseGeneratedAt(textOrNull(root, "generated_at"), summaryPath),
                textOrNull(root, "assumption"),
                textOrNull(root.path("scraper_inputs"), "courses_file"),
                textOrNull(root.path("scraper_inputs"), "skill_trends_file"),
                parseModelMetrics(root.path("model")),
                topPacks.isEmpty() ? null : topPacks.get(0),
                topPacks
        );
    }

    private Path resolveSummaryPath() {
        Path outputDir = resolveOutputDirectory();
        List<String> candidateFiles = List.of(
                packConversionSummaryFileName,
                "pack_conversion_summary.json"
        );

        for (String fileName : candidateFiles) {
            Path summaryPath = outputDir.resolve(fileName).normalize();
            if (Files.isRegularFile(summaryPath)) {
                return summaryPath;
            }
        }

        throw new BusinessRuleException("Pack conversion summary file not found in: " + outputDir);
    }

    private Path resolveOutputDirectory() {
        Path workingDir = Paths.get("").toAbsolutePath().normalize();
        List<Path> candidates = List.of(
                Paths.get(packConversionOutputDirectory),
                Paths.get(packConversionOutputDirectory).toAbsolutePath(),
                workingDir.resolve("..").resolve("Conversion packs").resolve("outputs"),
                workingDir.resolve("Conversion packs").resolve("outputs")
        );

        for (Path candidate : candidates) {
            Path normalized = candidate.toAbsolutePath().normalize();
            if (Files.isDirectory(normalized)) {
                return normalized;
            }
        }

        throw new BusinessRuleException("Pack conversion output directory not found. Checked: " + candidates);
    }

    private JsonNode readSummaryJson(Path summaryPath) {
        try {
            return objectMapper.readTree(Files.newInputStream(summaryPath));
        } catch (IOException ex) {
            throw new BusinessRuleException("Failed to read pack conversion summary: " + ex.getMessage());
        }
    }

    private Instant parseGeneratedAt(String value, Path sourcePath) {
        if (value == null || value.isBlank()) {
            return fileModifiedAt(sourcePath);
        }
        try {
            return Instant.parse(value);
        } catch (DateTimeParseException ignored) {
        }
        try {
            return LocalDateTime.parse(value).atZone(ZoneId.systemDefault()).toInstant();
        } catch (DateTimeParseException ignored) {
        }
        return fileModifiedAt(sourcePath);
    }

    private Instant fileModifiedAt(Path filePath) {
        try {
            return Files.getLastModifiedTime(filePath).toInstant();
        } catch (IOException ex) {
            return Instant.now();
        }
    }

    private PackConversionSummaryResponseDTO.ModelMetricsDTO parseModelMetrics(JsonNode node) {
        return new PackConversionSummaryResponseDTO.ModelMetricsDTO(
                textOrNull(node, "algorithm"),
                intOrNull(node, "random_seed"),
                intOrNull(node, "user_count"),
                intOrNull(node, "training_rows"),
                intOrNull(node, "test_rows"),
                doubleOrNull(node, "positive_rate"),
                doubleOrNull(node, "accuracy"),
                doubleOrNull(node, "roc_auc")
        );
    }

    private List<PackConversionSummaryResponseDTO.PackConversionScoreDTO> parseTopConversionPacks(JsonNode node, int limit) {
        List<PackConversionSummaryResponseDTO.PackConversionScoreDTO> items = new ArrayList<>();
        if (!node.isArray()) {
            return items;
        }

        int count = 0;
        for (JsonNode item : node) {
            if (count++ >= limit) {
                break;
            }
            items.add(new PackConversionSummaryResponseDTO.PackConversionScoreDTO(
                    intOrNull(item, "pack_id"),
                    textOrNull(item, "title"),
                    intOrNull(item, "category_id"),
                    textOrNull(item, "category_name"),
                    textOrNull(item, "level"),
                    textOrNull(item, "primary_skill"),
                    doubleOrNull(item, "sale_price"),
                    doubleOrNull(item, "original_price"),
                    doubleOrNull(item, "discount_pct"),
                    intOrNull(item, "duration_hours"),
                    doubleOrNull(item, "market_demand_score"),
                    doubleOrNull(item, "market_avg_roi"),
                    intOrNull(item, "market_course_count"),
                    doubleOrNull(item, "avg_views"),
                    doubleOrNull(item, "avg_cart_adds"),
                    doubleOrNull(item, "observed_purchase_rate"),
                    doubleOrNull(item, "predicted_conversion_rate"),
                    doubleOrNull(item, "expected_buyers"),
                    intOrNull(item, "training_support"),
                    doubleOrNull(item, "expected_revenue"),
                    intOrNull(item, "conversion_rank")
            ));
        }
        return items;
    }

    private String textOrNull(JsonNode node, String field) {
        JsonNode value = node.path(field);
        return value.isMissingNode() || value.isNull() ? null : value.asText(null);
    }

    private Double doubleOrNull(JsonNode node, String field) {
        JsonNode value = node.path(field);
        return value.isNumber() ? value.asDouble() : null;
    }

    private Integer intOrNull(JsonNode node, String field) {
        JsonNode value = node.path(field);
        return value.isInt() || value.isLong() ? value.asInt() : null;
    }
}

File: PIDEV-YBrainy-E-leraning-certifications-Platform/user/angular/angular-app/src/app/services/pack-conversion.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { catchError, map, Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { PackConversionSummary } from '../models/pack-conversion.model';

interface PackConversionAssetSummary {
  generated_at?: string | null;
  assumption?: string | null;
  scraper_inputs?: {
    courses_file?: string | null;
    skill_trends_file?: string | null;
  } | null;
  model?: {
    algorithm?: string | null;
    random_seed?: number | null;
    user_count?: number | null;
    training_rows?: number | null;
    test_rows?: number | null;
    positive_rate?: number | null;
    accuracy?: number | null;
    roc_auc?: number | null;
  } | null;
  top_conversion_packs?: Array<{
    pack_id?: number | null;
    title?: string | null;
    category_id?: number | null;
    category_name?: string | null;
    level?: string | null;
    primary_skill?: string | null;
    sale_price?: number | null;
    original_price?: number | null;
    discount_pct?: number | null;
    duration_hours?: number | null;
    market_demand_score?: number | null;
    market_avg_roi?: number | null;
    market_course_count?: number | null;
    avg_views?: number | null;
    avg_cart_adds?: number | null;
    observed_purchase_rate?: number | null;
    predicted_conversion_rate?: number | null;
    expected_buyers?: number | null;
    training_support?: number | null;
    expected_revenue?: number | null;
    conversion_rank?: number | null;
  }>;
}

@Injectable({
  providedIn: 'root'
})
export class PackConversionService {
  private readonly apiUrl = `${environment.apiUrl}/admin/packs/conversion`;
  private readonly fallbackSummaryUrl = 'assets/conversion-packs/pack_conversion_summary.json';

  constructor(private http: HttpClient) { }

  getSummary(limit = 10): Observable<PackConversionSummary> {
    const params = new HttpParams().set('limit', String(limit));
    return this.http.get<PackConversionSummary>(`${this.apiUrl}/summary`, { params }).pipe(
      catchError(() => this.getAssetSummary(limit))
    );
  }

  private getAssetSummary(limit: number): Observable<PackConversionSummary> {
    return this.http.get<PackConversionAssetSummary>(this.fallbackSummaryUrl).pipe(
      map((summary) => this.mapAssetSummary(summary, limit))
    );
  }

  private mapAssetSummary(summary: PackConversionAssetSummary, limit: number): PackConversionSummary {
    const topConversionPacks = (summary.top_conversion_packs ?? []).slice(0, limit).map((item) => ({
      packId: item.pack_id ?? null,
      title: item.title ?? null,
      categoryId: item.category_id ?? null,
      categoryName: item.category_name ?? null,
      level: item.level ?? null,
      primarySkill: item.primary_skill ?? null,
      salePrice: item.sale_price ?? null,
      originalPrice: item.original_price ?? null,
      discountPct: item.discount_pct ?? null,
      durationHours: item.duration_hours ?? null,
      marketDemandScore: item.market_demand_score ?? null,
      marketAvgRoi: item.market_avg_roi ?? null,
      marketCourseCount: item.market_course_count ?? null,
      avgViews: item.avg_views ?? null,
      avgCartAdds: item.avg_cart_adds ?? null,
      observedPurchaseRate: item.observed_purchase_rate ?? null,
      predictedConversionRate: item.predicted_conversion_rate ?? null,
      expectedBuyers: item.expected_buyers ?? null,
      trainingSupport: item.training_support ?? null,
      expectedRevenue: item.expected_revenue ?? null,
      conversionRank: item.conversion_rank ?? null
    }));

    return {
      sourceFile: this.fallbackSummaryUrl,
      generatedAt: summary.generated_at ?? new Date().toISOString(),
      assumption: summary.assumption ?? null,
      coursesFile: summary.scraper_inputs?.courses_file ?? null,
      skillTrendsFile: summary.scraper_inputs?.skill_trends_file ?? null,
      modelMetrics: summary.model ? {
        algorithm: summary.model.algorithm ?? null,
        randomSeed: summary.model.random_seed ?? null,
        userCount: summary.model.user_count ?? null,
        trainingRows: summary.model.training_rows ?? null,
        testRows: summary.model.test_rows ?? null,
        positiveRate: summary.model.positive_rate ?? null,
        accuracy: summary.model.accuracy ?? null,
        rocAuc: summary.model.roc_auc ?? null
      } : null,
      champion: topConversionPacks[0] ?? null,
      topConversionPacks
    };
  }
}
